File: event-bus-rabbit/src/main/java/pegasus/topology/service/CompositeTopologyManager.java
package pegasus.topology.service;

import java.util.PriorityQueue;

import pegasus.eventbus.amqp.RoutingInfo;
import pegasus.eventbus.client.EventManager;

public class CompositeTopologyManager implements TopologyManager {

    private PriorityQueue<TopoWrapper> topologyWrappers = new PriorityQueue<TopoWrapper>();

    @Override
    public RoutingInfo getRoutingInfoForEvent(Class<?> eventType) {
        RoutingInfo route = null;
        for (TopoWrapper topoWrapper : topologyWrappers) {
            route = topoWrapper.topologyManager.getRoutingInfoForEvent(eventType);
            if (route != null) {
                break;
            }
        }
        return route;
    }

    @Override
    public RoutingInfo[] getRoutingInfoForNamedEventSet(String eventSetName) {
        RoutingInfo[] routes = null;
        for (TopoWrapper topoWrapper : topologyWrappers) {
            routes = topoWrapper.topologyManager.getRoutingInfoForNamedEventSet(eventSetName);
            if (routes != null) {
                break;
            }
        }
        return routes;
    }

    public void addManager(TopologyManager topologyManager) {
        topologyWrappers.add(new TopoWrapper(topologyManager));
    }

    public void addManager(TopologyManager topologyManager, int priority) {
        topologyWrappers.add(new TopoWrapper(topologyManager, priority));
    }

    public void removeManager(TopologyManager topologyManager) {
        topologyWrappers.remove(topologyManager);
    }

    @Override
    public void start(EventManager eventManager) {
        for (TopoWrapper topoWrapper : topologyWrappers) {
            topoWrapper.topologyManager.start(eventManager);
        }
    }

    @Override
    public void stop() {
        for (TopoWrapper topoWrapper : topologyWrappers) {
            topoWrapper.topologyManager.stop();
        }
    }

    private class TopoWrapper implements Comparable<TopoWrapper> {

        private static final int DEFAULT_PRIORITY = 1;

        public int               priority         = DEFAULT_PRIORITY;
        public TopologyManager   topologyManager;

        public TopoWrapper(TopologyManager topologyManager) {
            this.topologyManager = topologyManager;
        }

        public TopoWrapper(TopologyManager topologyManager, int priority) {
            this.topologyManager = topologyManager;
            this.priority = priority;
        }

        @Override
        public int compareTo(TopoWrapper topoWrapper) {
            if (priority < topoWrapper.priority) {
                return -1;
            } else if (priority > topoWrapper.priority) {
                return 1;
            }

            return 0;
        }
    }

}

File: event-bus-rabbit/src/main/dotnet/topology/CompositeTopologyManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using log4net;

using pegasus.eventbus.client;
using pegasus.eventbus.amqp;


namespace pegasus.eventbus.topology
{
    public class CompositeTopologyManager : ITopologyService
    {
        private static readonly ILog LOG = LogManager.GetLogger(typeof(CompositeTopologyManager));

        private Queue<ITopologyService> _topologyManagers;


        public CompositeTopologyManager()
        {
            _topologyManagers = new Queue<ITopologyService>();
        }


        public void Append(ITopologyService topologyService)
        {
            LOG.DebugFormat("Appending topology service {0} to the composite topology service", topologyService.GetHashCode());

            _topologyManagers.Enqueue(topologyService);
        }

        public void Start(IEventManager eventManager)
        {
            LOG.Debug("Composite Topology Service is starting");

            _topologyManagers.ToList().ForEach(mgr => mgr.Start(eventManager));
        }

        public void Close()
        {
            LOG.Debug("Composite Topology Service is closing");

            _topologyManagers.ToList().ForEach(mgr => mgr.Close());
        }

        public RoutingInfo GetRoutingInfoForEventOfType(Type evType)
        {
            RoutingInfo route = null;

            _topologyManagers.FirstOrDefault(svc =>
            {
                route = svc.GetRoutingInfoForEventOfType(evType);
                return (null != route);
            });

            if (null != route)
                LOG.DebugFormat("Returning route {0}:{1} for event of type {2}", route.RoutingKey, route.Exchange, evType.FullName);
            else
                LOG.DebugFormat("No route for event of type {0}", evType.FullName);

            return route;
        }

        public IEnumerable<RoutingInfo> GetRoutingInfoForNamedEventSet(string eventSetName)
        {
            IEnumerable<RoutingInfo> routeSet = null;

            _topologyManagers.FirstOrDefault(svc =>
            {
                routeSet = svc.GetRoutingInfoForNamedEventSet(eventSetName);
                return (null != routeSet);
            });

            if (null != routeSet)
                LOG.DebugFormat("Returning {0} routes for event set {1}", routeSet.Count(), eventSetName);
            else
                LOG.DebugFormat("No routes for event set {0}", eventSetName);

            return routeSet;
        }
    }
}
